### PR TITLE
Enable ESA and fine tune AIO

### PR DESCRIPTION
**Pull request by luiscast84**
- Added ARM template for Container
- Code for Deploy ESA function
- PV, PVC, App for ESA
- JSON for hydra config
- Removed RP for AIO

### CODE REVIEW

File: azure_jumpstart_ag/artifacts/PowerShell/Bootstrap.ps1
Comment (dkirby-ms) [inline azure_jumpstart_ag/artifacts/PowerShell/Bootstrap.ps1:1]: I see the $spnObjectId parameter is being set at Bootstrap but i did not see a spot where you are passing this value into the CSE extension. This needs to be added to the host.bicep file under the Custom Script extension so that the $spnObjectId value is set

**Review by dkirby-ms — CHANGES_REQUESTED** [review]

File: azure_jumpstart_ag/artifacts/PowerShell/Bootstrap.ps1
Comment (luiscast84) [reply azure_jumpstart_ag/artifacts/PowerShell/Bootstrap.ps1:1]: I made all necessary changes and finished some parametrization for ESA.

**Review by luiscast84 — COMMENTED** [review]
There is a new PR over this one.

File: azure_jumpstart_ag/artifacts/PowerShell/AgLogonScript.ps1 [hunk 1]
```diff
@@ -1834,6 +1817,53 @@ function Deploy-AIO {
     }
 }
 
+function Deploy-ESA {
```
Comment (sebassem) [inline azure_jumpstart_ag/artifacts/PowerShell/AgLogonScript.ps1:1820]: Should have a loop to deploy on all clusters

**Review by sebassem — CHANGES_REQUESTED** [review]
Minor changes

File: azure_jumpstart_ag/artifacts/settings/ESA/pv.yaml [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+kind: PersistentVolume
+metadata:
+    ### CREATE A NAME HERE ###
+    name: esaPVName
+    ### USE A NAMESPACE THAT MATCHES YOUR INTENDED CONSUMING POD OR "DEFAULT" ###
+    namespace: namespace
+spec:
+    capacity:
+        ### This storage capacity value is not enforced at this layer. ###
+        storage: 10Gi
```
Comment (fcabrera23) [inline azure_jumpstart_ag/artifacts/settings/ESA/pv.yaml:11]: Should we make this configurable by user?

File: azure_jumpstart_ag/manufacturing/bicep/clientVm/clientVm.bicep [hunk 3]
```diff
@@ -16,6 +16,10 @@ param windowsOSVersion string = '2022-datacenter-g2'
 @description('Location for all resources')
 param location string = resourceGroup().location
 
+
+@description('Name of the storage account')
+param aioStorageAccountName string = 'aiostg${namingGuid}'
```
Comment (fcabrera23) [inline azure_jumpstart_ag/manufacturing/bicep/clientVm/clientVm.bicep:21]: Do we plan to create the storage account for the user?

File: azure_jumpstart_ag/artifacts/PowerShell/AgLogonScript.ps1 [hunk 4]
```diff
@@ -1834,6 +1817,53 @@ function Deploy-AIO {
...
+    (Get-Content $esaappYaml ) -replace 'namespace', $aioNamespace | Set-Content $esaappYaml
+    (Get-Content $esaappYaml ) -replace 'esaPVCName', $esaPVCName | Set-Content $esaappYaml
+
+    # Enable ESA extension on the Arc-enabled cluster
+    Write-Host "[$(Get-Date -Format t)] INFO: Enabling ESA on the Arc-enabled cluster" -ForegroundColor DarkGray
+    az k8s-extension create --resource-group $resourceGroup --cluster-name $arcClusterName --cluster-type connectedClusters --name hydraext --extension-type microsoft.edgestorageaccelerator --config-file $esapvJson
+    kubectl create secret generic -n $aioNamespace "$arcClusterName"-secret --from-literal=azurestorageaccountkey=$esaSecret --from-literal=azurestorageaccountname=$aioStorageAccountName
+    
+    Write-Host "[$(Get-Date -Format t)] INFO: Deploying PV on the Arc-enabled cluster" -ForegroundColor DarkGray
+    kubectl apply -f $esapvYaml
```
Comment (fcabrera23) [inline azure_jumpstart_ag/artifacts/PowerShell/AgLogonScript.ps1:1858]: Probably we need to check that the required storage for that PV is available in the node/cluster

File: azure_jumpstart_ag/artifacts/settings/ESA/config.json [hunk 5]
```diff
@@ -0,0 +1,5 @@
+{
+    "hydra.acstorController.enabled": false,
+    "hydra.highAvailability.disk.storageClass": "local-path",
+    "hydra.cachedStorageSize": "20Gi"
+}
```
Comment (fcabrera23) [inline azure_jumpstart_ag/artifacts/settings/ESA/config.json:5]: Is this file required for single-node clusters?

File: azure_jumpstart_ag/artifacts/settings/ESA/configPod.yaml [hunk 6]
```diff
@@ -0,0 +1,35 @@
+kind: Deployment
```
Comment (fcabrera23) [inline azure_jumpstart_ag/artifacts/settings/ESA/configPod.yaml:1]: What's the purpose of this configPod? This would be part of the end deployment, like the EdgeAI pod that users this PV through a PVC

File: azure_jumpstart_ag/artifacts/settings/ESA/pvc.yaml [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+            storage: 5Gi 
+    storageClassName: esa
+    volumeMode: Filesystem
+    ### THIS NAME IS FROM YOUR PV CONFIG ###
+    volumeName: esaPVName
+status: 
+    accessModes: 
+        - ReadWriteMany 
+    capacity: 
+        storage: 5Gi
```
Comment (fcabrera23) [inline azure_jumpstart_ag/artifacts/settings/ESA/pvc.yaml:22]: Any reason why are we keeping PV = 10GB and PVC to 5GB?

File: azure_jumpstart_ag/manufacturing/bicep/storage/storageAccount.bicep [hunk 8]
```diff
@@ -34,6 +37,13 @@ resource storageQueue 'Microsoft.Storage/storageAccounts/queueServices/queues@20
   name: storageQueueName
 }
 
+resource storageAccountName_default_container 'Microsoft.Storage/storageAccounts/blobServices/containers@2021-04-01' = {
+  name: '${storageAccountName}/default/${stcontainerName}'
+  dependsOn: [
+    storageAccount
+  ]
+}
+
```
Comment (fcabrera23) [inline azure_jumpstart_ag/manufacturing/bicep/storage/storageAccount.bicep:46]: nit: extra line

**Review by fcabrera23 — CHANGES_REQUESTED** [review]
Added a couple of nits and comments

File: azure_jumpstart_ag/artifacts/PowerShell/AgLogonScript.ps1 [hunk 9]
```diff
@@ -1834,6 +1817,53 @@ function Deploy-AIO {
     }
 }
 
+function Deploy-ESA {
+    # increase the maximum number of files
+    Invoke-AksEdgeNodeCommand -NodeType "Linux" -Command 'echo 'fs.inotify.max_user_instances = 1024' | sudo tee -a /etc/sysctl.conf && sudo sysctl -p'
```
Comment (sebassem) [inline azure_jumpstart_ag/artifacts/PowerShell/AgLogonScript.ps1:1822]: This command needs to run on both cluster VMs

Comment (fcabrera23) [reply azure_jumpstart_ag/artifacts/PowerShell/AgLogonScript.ps1:1822]: Should we just use a busybox container running as root for configuration, and then run it as a daemon set?

File: azure_jumpstart_ag/artifacts/PowerShell/AgLogonScript.ps1 (hunk 4 above)
Comment (luiscast84) [reply azure_jumpstart_ag/artifacts/PowerShell/AgLogonScript.ps1:1858]: Agree, shall I suggest getting this automation running OK end to end and add those checks later?

File: azure_jumpstart_ag/artifacts/PowerShell/AgLogonScript.ps1 [hunk 10]
```diff
@@ -1582,6 +1582,10 @@ function Deploy-InfluxDb {
     $influxdbYaml = "$aioToolsDir\influxdb.yml"
     $influxImportYaml = "$aioToolsDir\influxdb-import-dashboard.yml"
     $mqttExplorerSettings = "$aioToolsDir\mqtt_explorer_settings.json"
+    $esapvJson = "$aioToolsDir\config.json"
```
[the remark this answers is not in the document]
Comment (luiscast84) [reply azure_jumpstart_ag/artifacts/PowerShell/AgLogonScript.ps1:1585]: Yes, that's right. I put it all in the wrong place. Thanks

File: azure_jumpstart_ag/artifacts/PowerShell/AgLogonScript.ps1 (hunk 9 above)
Comment (luiscast84) [reply azure_jumpstart_ag/artifacts/PowerShell/AgLogonScript.ps1:1822]: There was a typo in the line

File: azure_jumpstart_ag/manufacturing/bicep/clientVm/clientVm.bicep (hunk 3 above)
Comment (luiscast84) [reply azure_jumpstart_ag/manufacturing/bicep/clientVm/clientVm.bicep:21]: Yes, as this is a pre-req for ESA

File: azure_jumpstart_ag/artifacts/settings/ESA/pv.yaml (hunk 2 above)
Comment (luiscast84) [reply azure_jumpstart_ag/artifacts/settings/ESA/pv.yaml:11]: It's parametrizable via automation now with placeholders

File: azure_jumpstart_ag/artifacts/settings/ESA/configPod.yaml (hunk 6 above)
Comment (luiscast84) [reply azure_jumpstart_ag/artifacts/settings/ESA/configPod.yaml:1]: Test app to check if ESA is working

File: azure_jumpstart_ag/manufacturing/bicep/clientVm/clientVm.bicep [hunk 11]
```diff
@@ -27,8 +31,11 @@ param subnetId string
 @description('Client id of the service principal')
 param spnClientId string
 
-//@description('Azure service principal object id')
-//param spnObjectId string
+@description('Azure service principal object id')
+param spnObjectId string
+
+@description('Azure service principal subscription id')
+param subscriptionId string
```
[the remark this answers is not in the document]
Comment (luiscast84) [reply azure_jumpstart_ag/manufacturing/bicep/clientVm/clientVm.bicep:38]: It's used in the logonscript for setting the sub context?

[the remark this answers is not in the document]
Comment (sebassem) [reply azure_jumpstart_ag/manufacturing/bicep/clientVm/clientVm.bicep:38]: no need, see the change I suggested below to get the subId automatically using env variables
